dockerfile-image-update/src/test/java/com/salesforce/dockerfileimageupdate/utils/RateLimiterTest.java

```diff
@@ -0,0 +1,139 @@
+                (new RateLimiterEvent(null, rateLimit, durationLimit, tokenAddingRate)));
+        Thread t1 = new Thread(futureTask);
+        long startTime = System.currentTimeMillis();
+        t1.start();
+        Thread.sleep(1000);
+        // the token should be consumed within first 3s leaving the thread in blocked state as it'd be waiting for
+        // next token to be available
+        assertEquals(t1.getState().toString(), "TIMED_WAITING");
+        t1.join();
+        long endTime = System.currentTimeMillis();
+        assertTrue((endTime - startTime) > (tokenAddingRate * 1_000L));
+        assertEquals(t1.getState().toString(), "TERMINATED");
+        assertEquals(t1.getState().toString(), "TERMINATED");
+        assertEquals(futureTask.get(), Optional.ofNullable(null));
+    }
+
+    /**
+     * Using Custom Clock to increment clock counter
+     * Instead of 60 seconds this should take < second to get completed
+     *
+     * @throws Exception
+     */
+    @Test(dataProvider = "rateLimitingData")
+    public void testingAgainstCustomClock(int rateLimit, int durationLimit, int tokenAddingRate) throws Exception {
+        RateLimiterEvent event = new RateLimiterEvent
+                (new CustomTimeMeter(), rateLimit, durationLimit, tokenAddingRate);
+        FutureTask futureTask = new FutureTask(new RateLimiterCallable(event));
+        Thread t1 = new Thread(futureTask);
+        long startTime = System.currentTimeMillis();
+        t1.start();
+        t1.join();
+        long endTime = System.currentTimeMillis();
+        assertTrue((endTime - startTime) < (tokenAddingRate * 1_000L));
+        assertEquals(event.getProcessedEventCount(), rateLimit);
+        assertEquals(t1.getState().toString(), "TERMINATED");
+        assertEquals(futureTask.get(), Optional.ofNullable(null));
+    }
+}
+
+
+class RateLimiterCallable implements Callable {
+    RateLimiterEvent rateLimiterEvent;
+
+    RateLimiterCallable(RateLimiterEvent rateLimiterEvent) {
+        this.rateLimiterEvent = rateLimiterEvent;
+    }
+
+    @Test
+    @Override
+    public Optional<Exception> call() {
+        try {
+            RateLimiter rateLimiter = new RateLimiter(rateLimiterEvent.ratelimit, Duration.ofSeconds(rateLimiterEvent.durationLimit),
+                    Duration.ofSeconds(rateLimiterEvent.tokenAddingRate), RateLimiterEvent.clock);
+            int counter = 0;
+            while (counter++ < rateLimiterEvent.ratelimit) {
+                rateLimiter.consume();
+                rateLimiterEvent.incrementEventProcessed();
+                assertEquals(Thread.currentThread().getState().toString(), "RUNNABLE");
+                if (RateLimiterEvent.clock != null) {
+                    RateLimiterEvent.clock.addSeconds(rateLimiterEvent.tokenAddingRate);
+                }
+            }
+            return Optional.ofNullable(null);
+        } catch (InterruptedException e) {
+            System.out.println("exception thrown");
```

Comment:
Why are we printing to STDOUT? This'll look out of context if you're debugging tests, especially if they run in parallel. In the worst case `log.error` or something like that - or delgate to the test caller to throw the exception via junit

Reply:
the class was copied from the other pr. fixed it